sources/web/datalab/authentication.ts

```diff
@@ -0,0 +1,135 @@
+ * the License.
+ */
+
+/// <reference path="../../../externs/ts/node/node.d.ts" />
+/// <reference path="../../../externs/ts/node/node-cache.d.ts" />
+/// <reference path="common.d.ts" />
+
+
+import http = require('http');
+import httpapi = require('./httpapi');
+import logging = require('./logging');
+import NodeCache = require("node-cache");
+
+/**
+ * The cache service for whitelisted user.
+ */
+var authCache: NodeCache = new NodeCache();
+
+/**
+ * The application settings instance.
+ */
+var appSettings: common.Settings;
+
+function getAccessTokenFromMetadataServer(cb: common.Callback<string>) {
+  var headers: common.Map<string> = {'X-Google-Metadata-Request': 'True'};
+  try {
+    var tokenHost: string = "metadata.google.internal";
+    if (appSettings.metadataHost) {
+      tokenHost = appSettings.metadataHost;
+    }
+    httpapi.get(tokenHost, '/computeMetadata/v1/instance/service-accounts/default/token',
+        null, null, headers, function(e: Error, data: any) {
+      if (e) {
+        logging.getLogger().error(e, 'Failed to get access token from metadata service.');
+        cb && cb(e, null);
+        return;
+      }
+      if (data && data.access_token) {
+        cb && cb(null, data.access_token);
+      }
+      else {
+        logging.getLogger().error('Failed to get access token metadata service response.');
+        cb && cb(Error('metadata server'), null);
+      }
+    });
+  }
+  catch (e) {
+    logging.getLogger().error(e, 'Failed to get access token from metadata service.');
+    cb && cb(e, null);
+  }
+}
+
+function lookupUserFromDatastore(userId: string, accessToken: string, cb: common.Callback<boolean>) {
+  if (authCache.get(userId)) {
+    cb(null, true);
+    return;
+  }
+  var postData: any = {
+    'keys': [
+      {
+        'path': [
+          {
+            'kind': 'DatalabUser',
+            'name': userId
+          }
+        ]
+      }
+    ]
+  };
+  var dataStorePath: string = '/datastore/v1beta2/datasets/' + appSettings.projectId + '/lookup';
+  try {
+    httpapi.posts('www.googleapis.com', dataStorePath, null, postData, accessToken, null, function(e: Error, data: any) {
+      if (e) {
+        logging.getLogger().error(e, 'Failed to query Datastore for authentication.');
+        cb && cb(e, null);
+        return;
+      }
+      var found: boolean = data && data.found && data.found.length > 0;
+      if (found) {
+        // Add user to cache and set expiration to 10 mintutes.
+        authCache.set(userId, true, 600);
+      }
+      cb && cb(null, found);
+    });
+  } 
+  catch (e) {
+    logging.getLogger().error(e, 'Failed to query Datastore for authentication.');
+    cb && cb(e, null);
+  }
+}
+
+/**
+ * Checks whether a given user has access to this instance of Datalab.
+ */
+export function checkUserAccess(userId: string, cb: common.Callback<boolean>) {
+  getAccessTokenFromMetadataServer(function(e: Error, accessToken: string) {
```

Comment:
I think it would be slightly better if we looked up cache here, so as to avoid doing the metadata lookup as well, and secondly have lookupUserFromDataStore only do datastore work.